### PR TITLE
[MWPW-152968] mWeb - Passing ECID to Branch.io banner - Implementation

### PR DESCRIPTION
**Pull request by drashti1712**
* When a user who has opted into both analytics and advertising cookies and is using Android device, their unique ECID is passed to Branch.

Resolves: [MWPW-152968](https://jira.corp.adobe.com/browse/MWPW-152968)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.live/drafts/drashti/milo-banner/branch-banner?martech=off
- After: https://branch-ecid--milo--adobecom.hlx.live/drafts/drashti/milo-banner/branch-banner

### CODE REVIEW

File: libs/blocks/mobile-app-banner/mobile-app-banner.js [hunk 1]
```diff
@@ -13,8 +13,19 @@ async function getKey(product) {
   return keyMatch[0]?.key;
 }
 
+async function getECID() {
```
Comment (narcis-radu) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:16]: We should get the ECID from `alloy`, not the cookie. Here's how we did it for Jarvis:
https://github.com/adobecom/milo/blob/stage/libs/features/jarvis-chat.js#L234

File: libs/blocks/mobile-app-banner/mobile-app-banner.js [hunk 2]
```diff
@@ -48,8 +59,18 @@ function branchInit(key) {
       0
     );
     const privacyConsent = window.adobePrivacy?.hasUserProvidedConsent();
+    const isAndroid = navigator.userAgent.includes('Android');
+    
+    const cookieGrp = window.adobePrivacy?.activeCookieGroups();
+    const performanceCookieConsent = cookieGrp.includes('C0002');
+    const advertisingCookieConsent = cookieGrp.includes('C0004');
+
+    if ( performanceCookieConsent && advertisingCookieConsent && isAndroid ) {
```
Comment (narcis-radu) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:68]: no need for `{`

File: libs/blocks/mobile-app-banner/mobile-app-banner.js [hunk 3]
```diff
@@ -48,8 +59,18 @@ function branchInit(key) {
...
+    branch.setBranchViewData({ data: { ecid: ecidVal }});
```
Comment (narcis-radu) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:69]: Is it ok to set `ecid: null` ? I think you might want to check the value first

**Review by narcis-radu — CHANGES_REQUESTED** [review]

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 1 above)
Comment (mokimo) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:16]: Could you re-use alloy to retrieve the ECID?
https://github.com/adobecom/milo/blob/stage/libs/blocks/global-navigation/global-navigation.js#L508

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 3 above)
Comment (mokimo) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:69]: There's `window.adobePrivacy` - however waiting might affect performance.

image

File: libs/blocks/mobile-app-banner/mobile-app-banner.js [hunk 4]
```diff
@@ -13,8 +13,19 @@ async function getKey(product) {
...
+async function getECID() {
+  let ecid = null;
+  const cookiesArr = document.cookie.split(';');
+  const regex = /^AMCV_[A-F0-9]+%40AdobeOrg=MCMID\|\d+$/;
+  cookiesArr.some((el) => {
+    if (regex.test(el.trim())) [, ecid] = el.split('MCMID|');
+    return ecid;
+  });
+  return ecid;
+}
```
Comment (mokimo) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:25]: You can also use `.find` to directly retrieve the matching element.
```suggestion
function getEcid() {
  const cookiesArr = document.cookie.split(';');
  const regex = /^AMCV_[A-F0-9]+%40AdobeOrg=MCMID\|\d+$/;
  const cookie = cookiesArr.find((el) => regex.test(el.trim()));
  return cookie ? cookie.split('MCMID|')[1] : null;
}
```

**Review by mokimo — APPROVED** [review]

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 2 above)
Comment (mokimo) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:68]: Does the linter not complain? 🤔 We don't leave spaces after `if` braces usually
```suggestion
    if (performanceCookieConsent && advertisingCookieConsent && isAndroid) {
```

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 1 above)
Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js:16]: Sure @narcis-radu , I made the suggested change. Thanks!

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 2 above)
Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js]: Done

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 3 above)
Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js]: If ecid is null, branch.setBranchViewData will automatically not set the ecid value. So no need to check if it is null.

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 1 above)
Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js:16]: YEs, made these changes. Thanks

File: libs/blocks/mobile-app-banner/mobile-app-banner.js (hunk 2 above)
Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js]: For some reason, linter wasn't working for me.. checked on that now. I have made the suggested changes. 
Thanks a lot!

File: libs/blocks/mobile-app-banner/mobile-app-banner.js [hunk 5]
```diff
@@ -13,8 +13,18 @@ async function getKey(product) {
   return keyMatch[0]?.key;
 }
 
+async function getECID() {
+  let ecid = null;
+  if (window.alloy) {
+    await window.alloy('getIdentity').then((data) => {
```
Comment (narcis-radu) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:19]: you should have a catch here in case the promise does not resolve

Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js:19]: Yes sure, added this!

File: libs/blocks/mobile-app-banner/mobile-app-banner.js [hunk 6]
```diff
@@ -13,8 +13,18 @@ async function getKey(product) {
...
+      ecid = data?.identity?.ECID;
+    }).catch((err) => window.lana.log(err));
```
Comment (narcis-radu) [inline libs/blocks/mobile-app-banner/mobile-app-banner.js:21]: I think we should categorize the error and provide the appropriate labels, e.g. https://github.com/adobecom/milo/blob/a856073ad5b4f0d72642be05e56e6aee9a06c4d8/libs/blocks/quiz-results/quiz-results.js#L23

Comment (drashti1712) [reply libs/blocks/mobile-app-banner/mobile-app-banner.js]: Sure, added the same.

**Review by narcis-radu — APPROVED** [review]

**Comment by spadmasa** [comment]
Validation done on the below flows in url :
https://branch-ecid--milo--adobecom.hlx.live/drafts/drashti/milo-banner/branch-banner
https://main--cc--adobecom.hlx.page/drafts/Souj/branch-io/branch-io-page11?milolibs=branch-ecid
details are added in https://jira.corp.adobe.com/browse/MWPW-152968

**Pull request merged**